src/compose/images.ts

```diff
@@ -0,0 +1,649 @@
+import * as validation from '../lib/validation';
+import Logger from '../logger';
+
+interface ImageEvents {
+	change: void;
+}
+
+type ImageEventEmitter = StrictEventEmitter<EventEmitter, ImageEvents>;
+
+interface ImageConstructOpts {
+	docker: DockerUtils;
+	logger: Logger;
+	db: Database;
+}
+
+interface FetchProgressEvent {
+	percentage: number;
+}
+
+// TODO: This is copied from src/lib/docker-utils.d.ts but because of the
+// export mechanism used, we can't export it. Once we convert docker-utils
+// to typescript, remove this
+interface DeltaFetchOptions {
+	deltaRequestTimeout: number;
+	deltaApplyTimeout: number;
+	deltaRetryCount: number;
+	deltaRetryInterval: number;
+	uuid: string;
+	currentApiKey: string;
+	deltaEndpoint: string;
+	apiEndpoint: string;
+	deltaSource: string;
+	deltaSourceId: string;
+	deltaVersion: string;
+}
+
+type FetchOptions = SchemaReturn<'fetchOptions'> & { deltaSource?: string };
+
+export interface Image {
+	id: number;
+	// image registry/repo@digest or registry/repo:tag
+	name: string;
+	appId: number;
+	serviceId: number;
+	serviceName: string;
+	// Id from balena api
+	imageId: number;
+	releaseId: number;
+	dependent: number;
+	dockerImageId: string;
+	status: 'Downloading' | 'Downloaded' | 'Deleting';
+	downloadProgress: Nullable<number>;
+}
+
+// TODO: This is necessary for the format() method, but I'm not sure
+// why, and it seems like a bad idea as it is. Fix the need for this.
+type MaybeImage = { [key in keyof Image]: Image[key] | null };
+
+// TODO: Remove the need for this type...
+type NormalisedDockerImage = Docker.ImageInfo & {
+	NormalisedRepoTags: string[];
+};
+
+export class Images extends (EventEmitter as {
+	new (): ImageEventEmitter;
+}) {
+	private docker: DockerUtils;
+	private logger: Logger;
+	private db: Database;
+
+	private imageCleanupFailures: Dictionary<number> = {};
+	// A store of volatile state for images (e.g. download progress), indexed by imageId
+	private volatileState: { [imageId: number]: Image } = {};
+
+	public constructor(opts: ImageConstructOpts) {
+		super();
+
+		this.docker = opts.docker;
+		this.logger = opts.logger;
+		this.db = opts.db;
+	}
+
+	public async triggerFetch(
+		image: Image,
+		opts: FetchOptions,
+		onFinish = _.noop,
+	): Promise<null> {
+		const onProgress = (progress: FetchProgressEvent) => {
+			// Only report the percentage if we haven't finished fetching
+			if (this.volatileState[image.imageId] != null) {
+				this.reportChange(image.imageId, {
+					downloadProgress: progress.percentage,
+				});
+			}
+		};
+
+		let success: boolean;
+		try {
+			const imageName = await this.normalise(image.name);
+			image = _.clone(image);
+			image.name = imageName;
+
+			await this.markAsSupervised(image);
+
+			const img = await this.inspectByName(image.name);
+			await this.db
+				.models('image')
+				.update({ dockerImageId: img.Id })
+				.where(image);
+
+			onFinish(true);
+			return null;
+		} catch (e) {
```

Comment:
If `e` is the error thrown by `await this.normalise(image.name)`, then it seems that the error is ignored? Maybe it's in front of my eyes, but I don't see where variable `e` is referenced...

Reply:
Ah this is where the confusion is coming from: https://github.com/balena-io/balena-supervisor/blob/master/src/compose/images.coffee#L59 the error is ignored here and assumed to be because the image does not exist. I'll add a specific check for this.